esp-hal-common/src/lib.rs

```diff
@@ -101,3 +102,118 @@ pub fn get_core() -> Cpu {
     #[cfg(feature = "single_core")]
     Cpu::ProCpu
 }
+
+mod critical_section_impl {
+    struct CriticalSection;
+
+    critical_section::set_impl!(CriticalSection);
+
+    #[cfg(target_arch = "xtensa")]
+    mod xtensa {
+
+        unsafe impl critical_section::Impl for super::CriticalSection {
+            unsafe fn acquire() -> critical_section::RawRestoreState {
+                let tkn: critical_section::RawRestoreState;
+                core::arch::asm!("rsil {0}, 15", out(reg) tkn);
+                #[cfg(feature = "multicore")]
+                {
+                    let guard = multicore::MULTICORE_LOCK.lock();
+                    core::mem::forget(guard); // forget it so drop doesn't run
+                }
+                tkn
+            }
+
+            unsafe fn release(token: critical_section::RawRestoreState) {
+                if token != 0 {
+                    #[cfg(feature = "multicore")]
+                    {
+                        debug_assert!(multicore::MULTICORE_LOCK.is_owned_by_current_thread());
+                        // safety: we logically own the mutex from acquire()
+                        multicore::MULTICORE_LOCK.force_unlock();
+                    }
+                    core::arch::asm!("wsr.ps {0}", in(reg) token)
```

Comment:
I believe it is important to add an `rsync` here to ensure the value is committed to the PS special register.